Add internal/intermediate struct for provenance data

mariaschett opened the pull request:
Adds and uses an internal intermediate struct `ProvenanceIR` in the verifier.

- [x] Tests pass
- [N/A] Appropriate changes to README are included in PR

rbehjati commented on internal/endorser/endorser.go:
```diff
@@ -37,7 +38,9 @@ import (
 // remote file (using the `http/https` scheme).
 func GenerateEndorsement(binaryDigest string, validityDuration amber.ClaimValidity, provenanceURIs []string) (*intoto.Statement, error) {
-	verifiedProvenances, err := loadAndVerifyProvenances(provenanceURIs, binaryDigest)
+	expected := verifier.ProvenanceIR{BinarySHA256Digest: []string{binaryDigest}}
```
How about instead passing `expected verifier.ProvenanceIR` as input to `GenerateEndorsement`? So that the caller of `GenerateEndorsement` is responsible for creating an instance of `verifier.ProvenanceIR`.

rbehjati commented on internal/endorser/endorser.go:
```diff
@@ -93,9 +96,11 @@ func loadAndVerifyProvenances(provenanceURIs []string, binaryDigest string) (*am
 // binary digest. An error is returned if any of the conditions above are violated.
 // TODO(b/222440937): Document any additional checks.
-func verifyProvenances(provenances []slsa.ValidatedProvenance, binaryDigest string) error {
+func verifyProvenances(provenances []slsa.ValidatedProvenance, expected verifier.ProvenanceIR) error {
```
```suggestion
func verifyProvenances(provenances []slsa.ValidatedProvenance, referenceValues verifier.ProvenanceIR) error {
```

Optional suggestion to be more specific :)

rbehjati commented on internal/verifier/verifier.go:
```diff
@@ -78,10 +78,16 @@ func (verifier *ReproducibleProvenanceVerifier) Verify(provenanceFilePath string
+	actual.FromAmber(provenance)
 
-	if err := buildConfig.VerifyBinarySHA256Digest(expectedBinaryDigest); err != nil {
+	expectedBinaryDigest, err := buildConfig.ComputeBinarySHA256Digest()
```
I don't quite see the benefit of this expansion. Can this logic be refactored to a `FromBuildConfig` function?

rbehjati commented on internal/verifier/verifier.go:
```diff
@@ -122,3 +128,47 @@ func (verifier *AmberProvenanceMetadataVerifier) Verify(provenanceFilePath strin
 	return nil
 }
+
+// Internal intermediate representation of Provenances for verification.
```
Golang comments start with the name of the element; in this case `ProvenanceIR`. The linters currently don't enforce it, but it is good to be consistent. Ref go/go-style/decisions#doc-comments.

rbehjati commented on internal/verifier/verifier.go:
```diff
@@ -122,3 +128,47 @@ func (verifier *AmberProvenanceMetadataVerifier) Verify(provenanceFilePath strin
+	BinarySHA256Digest []string
+}
+
+func (p *ProvenanceIR) FromSLSA(provenance *slsa.ValidatedProvenance) {
```
I am not sure what golang conventions are, but I think if `FromSLSA` and `FromAmber` are functions, instead of methods on `ProvenanceIR` the code will be nicer. In particular, you'd have:

```
actual := verifier.FromSLSA(&provenances[index])
```

instead of 

```
var actual verifier.ProvenanceIR
actual.FromSLSA(&provenances[index])
```

rbehjati reviewed (commented):
Nice. Thanks :) 

My main comments are on the logic of `VerifyWithReference`. Let me know if you'd rather discuss separately.

mariaschett commented on internal/verifier/verifier.go:
```diff
@@ -122,3 +128,47 @@ func (verifier *AmberProvenanceMetadataVerifier) Verify(provenanceFilePath strin
+	BinarySHA256Digest []string
+}
+
+func (p *ProvenanceIR) FromSLSA(provenance *slsa.ValidatedProvenance) {
```
Hmmm... I have been thinking, but I don't see a strong reason for them. 

The mapping should be obvious from the code. 

We could use unit tests to document our assumptions, e.g., that "*amber.ValidatedProvenance contains a SHA256 hash of a single subject", but that should actually be implied by the validation.

Do you see a good case?

rbehjati replied:
It will make refactoring easier. When the logic gets more complicated, and if we decide to make changes to the mapping, having tests will make it easier. But we can also add tests once there are more fields in ProvenanceIR.

mariaschett commented on internal/verifier/verifier.go:
```diff
@@ -122,3 +128,47 @@ func (verifier *AmberProvenanceMetadataVerifier) Verify(provenanceFilePath strin
+// VerifyWithReference verifies a provenance against a given reference.
+// We assume that we want to verify all fields the actual ProvenanceIR contains against the expected.
+// TODO(b/222440937): In future, also verify the details of the given provenance and the signature.
+func VerifyWithReference(actual ProvenanceIR, expected ProvenanceIR) error {
```
I'd like to use the `AmberProvenanceMetadataVerifier` in the next PR as wrapper for the `ProvenanceIRVerifier` in the endorser.

mariaschett commented on internal/verifier/verifier.go:
```diff
@@ -122,3 +128,47 @@ func (verifier *AmberProvenanceMetadataVerifier) Verify(provenanceFilePath strin
+	}
+
+	if actual.BinarySHA256Digest == nil {
+		// We don't want to verify the binary SHA256 digest.
```
I suggest the convention to only verify fields which are not empty in `Got`, i.e., all fields to be optional for flexibility in using the `ProvenanceIRVerifier`.

I agree that the digest should be present, but I think that should be caught by the `From` conversion/validation.

mariaschett commented on internal/verifier/verifier.go:
```diff
@@ -122,3 +128,47 @@ func (verifier *AmberProvenanceMetadataVerifier) Verify(provenanceFilePath strin
+		return nil
+	}
+
+	for _, expected := range expected.BinarySHA256Digest {
```
Added a comment, hope this clarifies :)

mariaschett reviewed (commented):
Thank you for comments.

rbehjati commented on internal/endorser/endorser.go:
```diff
@@ -26,18 +26,19 @@ import (
 )
 
-// GenerateEndorsement generates an endorsement statement for the given binary SHA256 digest, for
+// GenerateEndorsement generates an endorsement statement for the given reference, for
```
hmm... this does not sound right now. We don't generate an Endorsement for a reference. 

```suggestion
// GenerateEndorsement generates an endorsement statement for the binary SHA256 digest in the given reference, for
```

rbehjati commented on internal/endorser/endorser.go:
```diff
@@ -76,13 +77,13 @@ func loadAndVerifyProvenances(provenanceURIs []string, binaryDigest string) (*am
 	verifiedProvenances := amber.VerifiedProvenanceSet{
 		BinaryName:   provenances[0].GetBinaryName(),
-		BinaryDigest: binaryDigest,
+		BinaryDigest: reference.BinarySHA256Digests[0],
```
This does not sound right. If the reference can contain multiple values for `BinarySHA256Digests`, the value for `BinaryDigest` should not come from the reference, but from the provenances. We also need to make sure that if there are multiple provenances, they should all agree on the binary SHA256 digest.

rbehjati replied:
Yes. That makes sense. Thanks :)

rbehjati commented on internal/verifier/verifier.go:
```diff
@@ -119,3 +119,65 @@ func (verifier *AmberProvenanceMetadataVerifier) Verify() error {
+		BinarySHA256Digests: []string{provenance.GetBinarySHA256Digest()}}
+}
+
+// FromSLSAv0 maps data from a validated Amber provenance to ProvenanceIR.
```
```suggestion
// FromAmber maps data from a validated Amber provenance to ProvenanceIR.
```

rbehjati commented on internal/verifier/verifier.go:
```diff
@@ -119,3 +119,65 @@ func (verifier *AmberProvenanceMetadataVerifier) Verify() error {
+}
+
+// TODO(b/222440937): In future, also verify the details of the given provenance and the signature.
+// We only verify fields which are not empty in Got, all empty fields are ignored.
```
```suggestion
// Verify verifies an instance of ProvenanceIRVerifier by comparing its Got and Want fields. We only verify fields which are not empty in Got.
```

rbehjati commented on internal/verifier/verifier.go:
```diff
@@ -119,3 +119,65 @@ func (verifier *AmberProvenanceMetadataVerifier) Verify() error {
+// If a field in Got contains more than one value, we return an error.
+func (verifier *ProvenanceIRVerifier) Verify() error {
+	if len(verifier.Got.BinarySHA256Digests) != 1 {
+		return fmt.Errorf("got not exactly one actual binary SHA256 digest (%v)", verifier.Got.BinarySHA256Digests)
```
```suggestion
		return fmt.Errorf("provenance must have exactly one binary SHA256 digest value, got (%v)", len(verifier.Got.BinarySHA256Digests))
```

rbehjati commented on internal/verifier/verifier.go:
```diff
@@ -119,3 +119,65 @@ func (verifier *AmberProvenanceMetadataVerifier) Verify() error {
+	return verifier.Got.verifyBinarySHA256Digest(verifier.Want)
+}
+
+// verifyBinarySHA256Digest verifies that a single binary SHA256 is contained in the reference binary SHA256 digests.
```
```suggestion
// verifyBinarySHA256Digest verifies that the binary SHA256 in this provenance is contained in the given reference binary SHA256 digests (in want).
```

rbehjati commented on internal/endorser/endorser_test.go:
```diff
@@ -44,7 +45,10 @@ func TestGenerateEndorsement_SingleValidEndorsement(t *testing.T) {
 	tempURI := "file://" + tempPath
 	provenances := []string{tempURI}
-	statement, err := GenerateEndorsement(binaryHash, validity, provenances)
+	reference := verifier.ProvenanceIR{
```
Can you add a test with multiple values in `BinarySHA256Digests`?

mariaschett replied:
Yes, I'll do that!

rbehjati reviewed (commented):
Thanks for the fixes :) I think the logic for handling multiple reference values for the binary digest is still not quite right. Please see my comment on the changes in `loadAndVerifyProvenances`.

mariaschett reviewed (commented):
Thank you for the good review, Razieh! I'll add the tests on Tuesday.

rbehjati reviewed (approved):
Very nice. Thanks :)

mariaschett commented on internal/endorser/endorser_test.go:
```diff
@@ -70,12 +74,15 @@ func TestGenerateEndorsement_MultipleValidEndorsement(t *testing.T) {
 		t.Fatalf("Could not load provenance: %v", err)
 	}
-	tempPath2, err := copyToTemp("../../testdata/provenance.json")
+	tempPath2, err := copyToTemp("../../testdata/provenance_variant.json")
```
@rbehjati I think in this test we just over-wrote the `provenance.json` in temp before. I created a variant of `provenance.json`. Does this make sense?

rbehjati replied:
The previous approach was more flexible. Now we have an additional test file to maintain, and it does not add much value.  Is there any difference between `provenance.json` and `provenance_variant.json`?

mariaschett commented on internal/endorser/endorser_test.go:
```diff
@@ -92,13 +99,80 @@ func TestGenerateEndorsement_FailingSingleRemoteProvenanceEndorsement(t *testing
 	}
 }
 
+func TestGenerateEndorsement_MultipleNotVerifiedOrConsistent(t *testing.T) {
```
Added tests. Could you have a look @rbehjati ?

rbehjati replied:
Thanks. This test checks several scenarios. Since the scenarios do not require heavy set up, I prefer separate test cases, one for each scenario. That way when a test fails it will be easier to figure out which scenario failed.

mariaschett reviewed (commented):
Added tests.

rbehjati commented on internal/endorser/endorser_test.go:
```diff
@@ -92,13 +99,80 @@ func TestGenerateEndorsement_FailingSingleRemoteProvenanceEndorsement(t *testing
+			provenancePaths: []string{"file://" + tempPath1, "file://" + tempPath2},
+			want:            "do not contain the actual binary SHA256 digest", // "provenances are not consistent"
+		},
+		// Provenances do not contain reference binary SHA256 digest and are inconsistent.
```
Remove this commented block of code?

rbehjati commented on internal/endorser/endorser_test.go:
```diff
@@ -92,13 +99,80 @@ func TestGenerateEndorsement_FailingSingleRemoteProvenanceEndorsement(t *testing
+			provenancePaths: []string{"file://" + tempPath1, "file://" + tempPath1Variant},
+			want:            "do not contain the actual binary SHA256 digest",
+		},
+		// Provenances contain reference binary SHA256 digest, but are inconsistent.
```
Don't we need a scenario where there are multiple consistent provenances, and multiple reference values? 
That is the most realistic scenario. 
The two interesting scenarios are (1) where provenances are consistent, and the digest is contained in the reference values, and (2) the provenances are consistent, but the digest is not contained in the reference values.

mariaschett replied:
I think (1) we had already (`TestGenerateEndorsement_MultipleValidEndorsement`) and (2) is the first "// Provenances do not contain reference binary SHA256 digest, but are consistent."

rbehjati replied:
(1) is different from `TestGenerateEndorsement_MultipleValidEndorsement` in that in `TestGenerateEndorsement_MultipleValidEndorsement` we have only a single reference value; I am looking for a scenario with multiple reference values for the digest.

(2) is more generic than `"// Provenances do not contain reference binary SHA256 digest, but are consistent."`, in that (2) has multiple reference values. But actually, I was a bit confused by the wording of that comment. The current wording sounds like that the "provenances do not contain a SHA256 digest", which is a different negative test.

rbehjati commented on testdata/invalid_hash_provenance.json:
```diff
@@ -1 +1,43 @@
-{"_type":"https://in-toto.io/Statement/v0.1","predicateType":"https://slsa.dev/provenance/v0.2","subject":[{"name":"test.txt","digest":{"sha256":"1234"}}],"predicate":{"builder":{"id":""},"buildType":"https://github.com/project-oak/transparent-release/schema/amber-slsa-buildtype/v1/provenance.json","invocation":{"configSource":{}},"buildConfig":{"command":["cp","testdata/static.txt","test.txt"]," ...
\ No newline at end of file
+{
```
I think we needed provenances to be formatted according to canonical JSON, but if the code is not sensitive to it, it is probably not so important for now.

rbehjati reviewed (commented):
Thanks for adding the tests. I think there are some missing scenarios still.

rbehjati commented on internal/endorser/endorser_test.go:
```diff
@@ -92,13 +99,80 @@ func TestGenerateEndorsement_FailingSingleRemoteProvenanceEndorsement(t *testing
+		provenancePaths []string
+		want            string
+	}{
+		// Provenances do not contain reference binary SHA256 digest, but are consistent.
```
```suggestion
		// Provenances do not contain the given reference binary SHA256 digest value, but are consistent.
```

mariaschett commented on internal/endorser/endorser_test.go:
```diff
@@ -92,13 +99,80 @@ func TestGenerateEndorsement_FailingSingleRemoteProvenanceEndorsement(t *testing
+			provenancePaths: []string{"file://" + tempPath1, "file://" + tempPath1Variant},
+			want:            "do not contain the actual binary SHA256 digest",
+		},
+		// Provenances contain reference binary SHA256 digest, but are inconsistent.
```
Yes, this is a confusing comment! Sorry, changed.

rbehjati replied:
oh! The `1234` is the invalid one... sorry about that.

mariaschett commented:
I seperated the tests and I think I cover all interesting cases now:  `ConsistentNotVerified`, `InconsistentVerified`, and `InconsistentNotVerified`.

`ConsistentVerified` is covered by `MultipleValidEndorsement`. I don't think adding more reference values would add anything.

rbehjati commented:
> I seperated the tests and I think I cover all interesting cases now: `ConsistentNotVerified`, `InconsistentVerified`, and `InconsistentNotVerified`.
> 

Many thanks. This is easier for me understand :) 

> `ConsistentVerified` is covered by `MultipleValidEndorsement`. I don't think adding more reference values would add anything.

I think it does because. There was a bug earlier where the first element in the reference values (`reference.BinarySHA256Digests[0]`) was used as the digest of the binary in the endorsement. That is the scenario that I wanted to be checked and prevented. I don't see how that is covered in the existing tests.

mariaschett commented:
> That is the scenario that I wanted to be checked and prevented.

Very true, thank you! I added multiple reference values in `MultipleValidEndorsement` ---and refactored to simplify tests.

rbehjati commented on internal/endorser/endorser_test.go:
```diff
@@ -75,12 +72,17 @@ func TestGenerateEndorsement_MultipleValidEndorsement(t *testing.T) {
 
-	testutil.AssertEq(t, "binary hash", statement.Subject[0].Digest["sha256"], binaryHash)
+	testutil.AssertEq(t, "binary name", provenanceSet.BinaryName, binaryName)
+	testutil.AssertEq(t, "binary hash", provenanceSet.BinaryDigest, binaryHash)
```
Nice assertion :)

rbehjati reviewed (approved):
Perfect. Thanks. Nice refactoring :)
